lib/deployServiceClient.js

```diff
@@ -0,0 +1,152 @@
+    directoryFilter: directoriesToIgnore
+  })
+    .on("data", entry => {
+      var fileName = `${assetsDirectory}/${entry.path}`;
+      assets.push({
+        fileName: fileName,
+        contentLength: fs.statSync(fileName)["size"],
+        contentType: mime.getType(fileName)
+      });
+    })
+    .on("end", () => socket.emit("deploy:presign", { urls: assets }));
+};
+
+const initializeConnection = () => {
+  return new Promise((resolve, reject) => {
+    try {
+      var serviceUrl = new url.URL(process.env.MARKETPLACE_URL);
+
+      if (!process.env.DEPLOY_SERVICE_URL) {
+        // Make it possible to set DEPLOY_SERVICE_URL without
+        // depending on Load Balancer settings
+        serviceUrl = process.env.DEPLOY_SERVICE_URL;
+      } else {
+        // We have one deploy-service per environent hosted at URI /deploy-service.*
+        serviceUrl.pathname = "/deploy-service";
+        serviceUrl = serviceUrl.href;
+      }
+      logger.Info(`Connecting to ${serviceUrl}`);
+      return resolve(
+        io.connect(
+          serviceUrl,
+          {
+            extraHeaders: {
+              token: process.env.MARKETPLACE_TOKEN,
+              marketplace_domain: url.parse(process.env.MARKETPLACE_URL)
+                .hostname
+            }
+          }
+        )
+      );
+    } catch (e) {
+      reject(e);
+    }
+  });
+};
+
+const presignUrlsAndUploadFiles = () => {
+  return new Promise((resolve, reject) => {
+    var assets = [];
+    var uploaded = {};
+
+    initializeConnection().then(
+      socket => {
+        socket.on("connect_error", err =>
+          reject(`[DeployService] Connection atempt failed ${err}`)
+        );
+
+        socket.on("connect_timeout", () =>
+          reject("[DeployService] Connection atempt failed (timeout)")
+        );
+
+        socket.on("error", err => reject(err));
+
+        socket.on("connect", () => {
+          logger.Info("[DeployService] Connected requesting presigned S3 urls");
+          sendRequestsForPresignedUrls(socket, assets);
+        });
+
+        socket.on("deploy:url", data => {
+          uploadFile(data.fileName, data.url, data.accessUrl)
+            .then(
+              () => {
+                logger.Print(".");
+                uploaded[data.fileName] = data.accessUrl;
+                assets = assets.filter(i => i.fileName !== data.fileName);
+                if (assets.length == 0) {
+                  logger.Print("\n");
+                  uploadManifest(uploaded).then(
+                    () => {
+                      setTimeout(() => {
+                        socket.disconnect();
+                        return resolve(true);
+                      }, 10000);
+                    },
+                    err => reject(err)
+                  );
+                }
+              },
+              err => reject(err)
+            )
+            .catch(err => reject(err));
+        });
+      },
+      err => reject(err)
```

Comment:
reject

Reply:
Explanation for all of those:

If function expect a function as an argument and all you do is run anotherFunction with the arguments from it, you can pass reference to the anotherFunction - it will pass all the arguments from above.

IE.

`$.get(url).then(console.log)` will pipe `then` result value to `console.log` function
or
["a", 5].map(isNaN) => [true, false]

Reply:
There are probably better explanation of that feature and my terminology is wrong ;)